core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -79,6 +88,44 @@ public static CloseableIterable<FileScanTask> splitFiles(
     return CloseableIterable.combine(splitTasks, tasks);
   }
 
+  /**
+   * Produces {@link CombinedScanTask combined tasks} from an iterable of {@link FileScanTask file
+   * tasks}, using an adaptive target split size that targets a minimum number of tasks
+   * (parallelism).
+   *
+   * @param files incoming iterable of file tasks
+   * @param parallelism target minimum number of tasks
+   * @param splitSize target split size
+   * @param lookback bin packing lookback
+   * @param openFileCost minimum file cost
+   * @return an iterable of combined tasks
+   */
+  public static CloseableIterable<CombinedScanTask> planTasksAdaptive(
```

Comment:
If we like this approach, the new "adaptive" planning methods would be called in the scan code instead of splitting and combining. That would require adding the parallelism option to our scan API, so I'd do it only for the new `BatchScan` API.